pkg/endpoint/restore.go

```diff
@@ -61,36 +60,14 @@ func ReadEPsFromDirNames(ctx context.Context, owner regeneration.Owner, basePath
 	possibleEPs := map[uint16]*Endpoint{}
 	for _, epDirName := range completeEPDirNames {
 		epDir := filepath.Join(basePath, epDirName)
-		readDir := func() string {
-			scopedLog := log.WithFields(logrus.Fields{
-				logfields.EndpointID: epDirName,
-				logfields.Path:       filepath.Join(epDir, common.CHeaderFileName),
-			})
-			scopedLog.Debug("Reading directory")
-			epFiles, err := ioutil.ReadDir(epDir)
-			if err != nil {
-				scopedLog.WithError(err).Warn("Error while reading directory. Ignoring it...")
-				return ""
-			}
-			cHeaderFile := common.FindEPConfigCHeader(epDir, epFiles)
-			if cHeaderFile == "" {
-				return ""
-			}
-			return cHeaderFile
-		}
-		// There's an odd issue where the first read dir doesn't work.
-		cHeaderFile := readDir()
-		if cHeaderFile == "" {
-			cHeaderFile = readDir()
-		}
-
+		cHeaderFile := filepath.Join(epDir, common.CHeaderFileName)
 		scopedLog := log.WithFields(logrus.Fields{
 			logfields.EndpointID: epDirName,
 			logfields.Path:       cHeaderFile,
 		})
 
-		if cHeaderFile == "" {
-			scopedLog.Warning("C header file not found. Ignoring endpoint")
+		if _, err := os.Stat(cHeaderFile); err != nil {
```

Comment:
Can we attempt 3 times with a 100 ms sleep in between before failing? If this file is not found the endpoint is not restored.

Reply:
^^^ If we think this actually does something meaningful, it'd be nice to get evidence somehow. Could we add a metric (not necessarily promethus, but something we can fetch via `cilium metrics list`) that this actually occurred?

That way over time we can check this and see if we really need it or if the reason we had to put this in in the first place has gone away.

Reply:
I agree that we need a way to get rid of this in the long term. It's either not an issue anymore or we need to reproduce and fix the underlying problem, but we don't want to carry this *ad aeternam* because we noticed it once.

Do we actually check metrics though? I was thinking we could print a debug message if the output changes with the 2nd/3rd calls and add that debug message to `badLogMessages`. That way we're more likely to notice this issue happening.

@aanm Do you remember in what context this was happening?

Reply:
>I agree that we need a way to get rid of this in the long term. It's either not an issue anymore or we need to reproduce and fix the underlying problem, but we don't want to carry this ad aeternam because we noticed it once.

I agree but the problem is that if the file is not found the endpoint, and connectivity, of the pod will be broken.

>Do we actually check metrics though? I was thinking we could print a debug message if the output changes with the 2nd/3rd calls and add that debug message to badLogMessages. That way we're more likely to notice this issue happening.

We do in the sysdumps. We can also add a `Warning` message with the string "BUG".

>@aanm Do you remember in what context this was happening?

This happened to me when I was developing the restore code. The endpoints were not being restored and I added debug messages all over the place until I got the "File not found" which wouldn't fail on the 2nd attempt.

Reply:
I pushed an update with 2 retries spaced by 100ms. I also added the BUG warning and a `badLogMessages` entry. I've run the `K8sUpdates` test 34 times locally to ensure the bug isn't actually too frequent, so that we won't block all PRs just after merging this.